## Summary (plain English)

This notebook tries to answer one question: is ASML's stock priced fairly today?

**What we expect ASML to earn going forward:** Sales are expected to jump sharply in 2026 (up
around 35%) as demand for AI-related chipmaking equipment stays strong, then grow more unevenly
after that — a slower year in 2028, then picking back up again — reaching around EUR 62 billion
in sales by 2030, up from about EUR 33 billion in 2025.

**How risky is holding ASML stock or debt:** Every investment needs to "clear a bar" — a minimum
return investors expect for the risk they're taking. For ASML, that bar comes out to about 9.3% a
year. Because ASML barely borrows any money (it's funded almost entirely by shareholders), this
number is really just what shareholders themselves expect to earn for the risk of owning the
stock.

**Two different ways of estimating what the stock is worth:**
1. The "ASML eventually becomes an ordinary company" view: if ASML's growth eventually settles
   down to match the pace of the wider economy, forever, the stock would be worth about EUR 593 a
   share — far below where it trades today.
2. The "ASML keeps being treated like a top-tier grower" view: if ASML keeps being valued the way
   the market has valued it, on average, over the last ten years, the stock would be worth about
   EUR 1,459 a share — very close to where it trades today.

**Bottom line:** ASML's actual share price today (around EUR 1,407) sits almost exactly where the
second view lands, and far above the first. That tells us the market isn't pricing ASML as an
"average" business — it's betting that ASML keeps growing strongly for years to come, the way it
always has. Whether the stock is a good buy from here depends almost entirely on whether you
believe that bet is right, not on any small tweak to the numbers in this model.

**A caution:** every number here comes from a real, sourced assumption (company guidance, industry
forecasts, market data), but it's still a model of the future, not a fact about it. Small changes
in the assumptions can move the answer a lot — see the sensitivity tables near the end for how
much.

# ASML Holding N.V. — DCF Valuation Model

## 1. Historical Financials (FY2021–FY2025)

**Reporting currency: EUR (€), in millions**, unless stated otherwise. ASML reports under US GAAP in
its quarterly/annual results press releases (it also files an IFRS-based Dutch statutory annual report,
but the quarterly press releases and the SEC 20-F use US GAAP). All figures below are taken from the
official **"Financial statements US GAAP" PDF exhibits** that ASML links from each Q4 press release —
these contain the full "Summary US GAAP Consolidated Statements of Operations" table, which is not
reproduced in the press release body text itself. Sources:

- FY2021: ASML Q4 & FY2021 press release (https://www.asml.com/en/news/press-releases/2022/q4-and-full-year-2021-financial-results),
  linked exhibit *"Financial statements US GAAP Q4 2021"*, page 1, "Summary US GAAP Consolidated
  Statements of Operations", column "Twelve months ended Dec 31, 2021".
- FY2022: ASML Q4 2022 press release (https://www.asml.com/en/news/press-releases/2023/q4-2022-financial-results),
  linked exhibit *"Financial statements US GAAP Q4 2022"*, page 1, "Summary US GAAP Consolidated
  Statements of Operations", column "Twelve months ended Dec 31, 2022".
- FY2023: ASML Q4 2023 press release (https://www.asml.com/en/news/press-releases/2024/q4-2023-financial-results),
  linked exhibit *"Financial statements US GAAP Q4 2023"*, page 1, "Summary US GAAP Consolidated
  Statements of Operations", column "Twelve months ended Dec 31, 2023".
- FY2024: ASML Q4 2024 press release (https://www.asml.com/en/news/press-releases/2025/q4-2024-financial-results),
  linked exhibit *"Financial statements US GAAP Q4 2024"*, page 1, "Summary US GAAP Consolidated
  Statements of Operations", column "Twelve months ended Dec 31, 2024".
- FY2025: ASML Q4 2025 press release (https://www.asml.com/en/news/press-releases/2026/q4-2025-financial-results),
  linked exhibit *"Financial statements US GAAP Q4 2025"*, page 3, "Summary US GAAP Consolidated
  Statements of Operations", column "Twelve months ended Dec 31, 2025".

These statements are explicitly labelled **unaudited** by ASML (they are interim summary statements
published alongside quarterly results, ahead of the audited Annual Report).

**Note on FY2021 "Other income":** ASML's FY2021 statement includes a line between gross profit and
R&D costs, "Other income" (€213.7m) — footnoted as the gain on the sale of ASML's Berliner Glas
subsidiaries. This line is not present in any other year's statement in this table, so it is added
as its own row (0 for all other years) rather than silently folded into EBIT, so that
Gross profit + Other income + R&D expenses + SG&A expenses reconciles exactly to Operating income
(EBIT) for every year — verified by an assertion in the code cell below.

**Note on "Profit before tax" and "Income tax expense":** these two rows come from further down the
*same* "Summary US GAAP Consolidated Statements of Operations" table already cited above for each
year (ASML's own labels are "Income before income taxes" and "Income tax expense") — no additional
source is needed. Both are fully disclosed for all five years. **Effective tax rate %** is not a
separate disclosed line; it's calculated in the code cell below as
Income tax expense ÷ Profit before tax, and will later anchor the tax-rate assumption in Section 5.
Note that Profit before tax also nets in "Interest and other, net" (not itself pulled into this
table), so Profit before tax will not exactly equal Operating income (EBIT) — that's expected, not
an error.

### What each line item means

- **Net sales** — total revenue for the year: the sum of net system sales (new lithography/metrology
  tools) and net service and field option sales (maintenance, upgrades, spares for the installed base).
- **Cost of sales** — the direct cost of producing and delivering those systems and services (materials,
  manufacturing labor, warranty, etc.).
- **Gross profit** — net sales minus cost of sales; what's left to cover R&D, SG&A, and generate profit.
- **Other income** — a non-recurring item ASML reported only in FY2021 (gain on the sale of its
  Berliner Glas subsidiaries); zero in every other year in this table. Included as its own row so the
  table reconciles exactly rather than leaving an unexplained gap in FY2021's EBIT.
- **R&D expenses** — spending on research and development of future lithography systems (EUV, High-NA
  EUV, etc.) — a large, strategically important cost for ASML given multi-year technology roadmaps.
- **SG&A expenses** — selling, general and administrative costs: sales force, corporate overhead,
  administration.
- **Operating income (EBIT)** — gross profit plus other income, minus R&D and SG&A; profit from core
  operations before interest and taxes. ASML calls this "Income from operations".
- **Profit before tax** — EBIT adjusted for non-operating items (net interest income/expense and
  other) — ASML's "Income before income taxes". The step between EBIT and net income that tax gets
  applied to.
- **Income tax expense** — corporate income tax ASML recognizes on profit before tax for the year.
- **Effective tax rate %** — Income tax expense ÷ Profit before tax; the *actual* average tax rate
  paid, which can differ from any single jurisdiction's statutory rate due to ASML's mix of
  international operations. Calculated, not disclosed as its own line.
- **Net income** — bottom-line profit after interest, taxes, and equity-method investment income —
  attributable to shareholders.
- **EPS (basic / diluted)** — net income divided by the weighted-average number of ordinary shares
  outstanding (diluted includes the dilutive effect of share-based awards).

We are **not** calculating margins or growth rates yet — just laying out the raw historicals cleanly.

In [1]:
import pandas as pd

# Figures in EUR millions, except EPS which is in EUR per share.
# Sign convention: cost/expense lines are stored as negative numbers (as ASML presents them),
# so each column sums down correctly:
#   Net sales + Cost of sales = Gross profit
#   Gross profit + Other income + R&D expenses + SG&A expenses = Operating income (EBIT)

historical_data = {
    "FY2021": {
        # Source: ASML Q4 & FY2021 press release -> linked exhibit "Financial statements US GAAP Q4 2021"
        # (https://www.asml.com/en/news/press-releases/2022/q4-and-full-year-2021-financial-results), p.1,
        # "Summary US GAAP Consolidated Statements of Operations", "Twelve months ended Dec 31, 2021"
        "Net sales": 18611.0,
        "Cost of sales": -8802.0,
        "Gross profit": 9809.0,
        # ASML reports a one-off "Other income" line (footnoted: gain on sale of Berliner Glas
        # subsidiaries) between Gross profit and R&D costs in FY2021 only - not present in any
        # other year in this table. Same source as above, p.1.
        "Other income": 213.7,
        "R&D expenses": -2547.0,
        "SG&A expenses": -725.6,
        "Operating income (EBIT)": 6750.1,   # ASML label: "Income from operations"
        "Profit before tax": 6705.5,         # ASML label: "Income before income taxes", same source p.1
        "Income tax expense": -1021.4,       # same source p.1
        "Net income": 5883.2,
        "EPS - basic (EUR)": 14.36,
        "EPS - diluted (EUR)": 14.34,
    },
    "FY2022": {
        # Source: ASML Q4 2022 press release -> linked exhibit "Financial statements US GAAP Q4 2022"
        # (https://www.asml.com/en/news/press-releases/2023/q4-2022-financial-results), p.1,
        # "Summary US GAAP Consolidated Statements of Operations", "Twelve months ended Dec 31, 2022"
        "Net sales": 21173.4,
        "Cost of sales": -10473.3,
        "Gross profit": 10700.1,
        "Other income": 0.0,  # no separate "Other income" line reported for FY2022 in the source
        "R&D expenses": -3253.5,
        "SG&A expenses": -945.9,
        "Operating income (EBIT)": 6500.7,
        "Profit before tax": 6456.1,         # ASML label: "Income before income taxes", same source p.1
        "Income tax expense": -969.9,        # same source p.1
        "Net income": 5624.2,
        "EPS - basic (EUR)": 14.14,
        "EPS - diluted (EUR)": 14.13,
    },
    "FY2023": {
        # Source: ASML Q4 2023 press release -> linked exhibit "Financial statements US GAAP Q4 2023"
        # (https://www.asml.com/en/news/press-releases/2024/q4-2023-financial-results), p.1,
        # "Summary US GAAP Consolidated Statements of Operations", "Twelve months ended Dec 31, 2023"
        "Net sales": 27558.5,
        "Cost of sales": -13422.4,
        "Gross profit": 14136.1,
        "Other income": 0.0,  # no separate "Other income" line reported for FY2023 in the source
        "R&D expenses": -3980.6,
        "SG&A expenses": -1113.2,
        "Operating income (EBIT)": 9042.3,   # ASML label: "Income from operations"
        "Profit before tax": 9083.5,         # ASML label: "Income before income taxes", same source p.1
        "Income tax expense": -1435.8,       # same source p.1
        "Net income": 7839.0,
        "EPS - basic (EUR)": 19.91,
        "EPS - diluted (EUR)": 19.89,
    },
    "FY2024": {
        # Source: ASML Q4 2024 press release -> linked exhibit "Financial statements US GAAP Q4 2024"
        # (https://www.asml.com/en/news/press-releases/2025/q4-2024-financial-results), p.1,
        # "Summary US GAAP Consolidated Statements of Operations", "Twelve months ended Dec 31, 2024"
        "Net sales": 28262.9,
        "Cost of sales": -13770.9,
        "Gross profit": 14492.0,
        "Other income": 0.0,  # no separate "Other income" line reported for FY2024 in the source
        "R&D expenses": -4303.7,
        "SG&A expenses": -1165.7,
        "Operating income (EBIT)": 9022.6,
        "Profit before tax": 9042.4,         # ASML label: "Income before income taxes", same source p.1
        "Income tax expense": -1680.6,       # same source p.1
        "Net income": 7571.6,
        "EPS - basic (EUR)": 19.25,
        "EPS - diluted (EUR)": 19.24,
    },
    "FY2025": {
        # Source: ASML Q4 2025 press release -> linked exhibit "Financial statements US GAAP Q4 2025"
        # (https://www.asml.com/en/news/press-releases/2026/q4-2025-financial-results), p.3,
        # "Summary US GAAP Consolidated Statements of Operations", "Twelve months ended Dec 31, 2025"
        "Net sales": 32667.3,
        "Cost of sales": -15409.3,
        "Gross profit": 17258.0,
        "Other income": 0.0,  # no separate "Other income" line reported for FY2025 in the source
        "R&D expenses": -4698.8,
        "SG&A expenses": -1257.8,
        "Operating income (EBIT)": 11301.4,
        "Profit before tax": 11406.1,        # ASML label: "Income before income taxes", same source p.3
        "Income tax expense": -2013.4,       # same source p.3
        "Net income": 9609.4,
        "EPS - basic (EUR)": 24.73,
        "EPS - diluted (EUR)": 24.71,
    },
}

historicals = pd.DataFrame(historical_data)

# Sanity check: Gross profit + Other income + R&D + SG&A should equal Operating income (EBIT)
# for every year, confirming nothing was mistranscribed.
computed_ebit = (
    historicals.loc["Gross profit"]
    + historicals.loc["Other income"]
    + historicals.loc["R&D expenses"]
    + historicals.loc["SG&A expenses"]
)
assert (computed_ebit - historicals.loc["Operating income (EBIT)"]).abs().max() < 1e-6, "EBIT does not reconcile"

# Effective tax rate is not a disclosed line - it's calculated from the two hardcoded rows above,
# rounded to 2dp for a clean display, and inserted directly beneath "Income tax expense".
effective_tax_rate = (historicals.loc["Income tax expense"].abs() / historicals.loc["Profit before tax"] * 100).round(2)
row_order = historicals.index.tolist()
row_order.insert(row_order.index("Income tax expense") + 1, "Effective tax rate %")
historicals.loc["Effective tax rate %"] = effective_tax_rate
historicals = historicals.loc[row_order]

historicals

,FY2021,FY2022,FY2023,FY2024,FY2025
Net sales,18611.00,21173.40,27558.50,28262.90,32667.30
Cost of sales,-8802.00,-10473.30,-13422.40,-13770.90,-15409.30
Gross profit,9809.00,10700.10,14136.10,14492.00,17258.00
Other income,213.70,0.00,0.00,0.00,0.00
R&D expenses,-2547.00,-3253.50,-3980.60,-4303.70,-4698.80
SG&A expenses,-725.60,-945.90,-1113.20,-1165.70,-1257.80
Operating income (EBIT),6750.10,6500.70,9042.30,9022.60,11301.40
Profit before tax,6705.50,6456.10,9083.50,9042.40,11406.10
Income tax expense,-1021.40,-969.90,-1435.80,-1680.60,-2013.40
Effective tax rate %,15.23,15.02,15.81,18.59,17.65


## 2. Margins & Growth

Everything in this section is **calculated from the "1. Historical Financials" DataFrame above** —
no new hardcoded numbers are introduced.

- **Net sales YoY growth %** — this year's net sales divided by last year's, minus one. It measures
  how fast the top line is growing *period over period*, which is the natural way to look at growth:
  there's no FY2020 figure in our table, so FY2021 has nothing to compare against and is left blank
  (NaN).
- **Gross margin %**, **R&D as % of net sales**, **SG&A as % of net sales**, **Operating margin %**,
  and **Net margin %** are all *common-sized* figures: each line item is expressed as a percentage of
  that same year's net sales, rather than as a year-over-year change. Common-sizing is the right lens
  for margins because it strips out the effect of the business simply being bigger or smaller in a
  given year — it answers "of every euro of revenue, how much falls through to gross profit / R&D /
  operating income / net income?", which is comparable across years (or against other companies)
  regardless of absolute scale.

In short: **growth rates compare a line item to itself across time**, while **margins compare a line
item to revenue within the same year**. Both are needed — growth tells you the trajectory, margins
tell you the profitability and cost structure driving that trajectory.

In [2]:
net_sales = historicals.loc["Net sales"]

margins_growth = pd.DataFrame(index=[
    "Net sales YoY growth %",
    "Gross margin %",
    "R&D as % of net sales",
    "SG&A as % of net sales",
    "Operating margin %",
    "Net margin %",
], columns=historicals.columns, dtype=float)

margins_growth.loc["Net sales YoY growth %"] = net_sales.pct_change() * 100
margins_growth.loc["Gross margin %"] = historicals.loc["Gross profit"] / net_sales * 100
margins_growth.loc["R&D as % of net sales"] = historicals.loc["R&D expenses"].abs() / net_sales * 100
margins_growth.loc["SG&A as % of net sales"] = historicals.loc["SG&A expenses"].abs() / net_sales * 100
margins_growth.loc["Operating margin %"] = historicals.loc["Operating income (EBIT)"] / net_sales * 100
margins_growth.loc["Net margin %"] = historicals.loc["Net income"] / net_sales * 100

# Formatted view for display; margins_growth itself keeps the underlying numeric values.
margins_growth_display = margins_growth.map(lambda v: f"{v:.1f}%" if pd.notna(v) else "–")
display(margins_growth_display)

# 5-year revenue CAGR, FY2021 -> FY2025 (4 periods of growth between 5 data points)
years_elapsed = 4
revenue_cagr = (net_sales["FY2025"] / net_sales["FY2021"]) ** (1 / years_elapsed) - 1
print(f"FY2021-FY2025 revenue CAGR: {revenue_cagr:.1%}")

,FY2021,FY2022,FY2023,FY2024,FY2025
Net sales YoY growth %,–,13.8%,30.2%,2.6%,15.6%
Gross margin %,52.7%,50.5%,51.3%,51.3%,52.8%
R&D as % of net sales,13.7%,15.4%,14.4%,15.2%,14.4%
SG&A as % of net sales,3.9%,4.5%,4.0%,4.1%,3.9%
Operating margin %,36.3%,30.7%,32.8%,31.9%,34.6%
Net margin %,31.6%,26.6%,28.4%,26.8%,29.4%


FY2021-FY2025 revenue CAGR: 15.1%


## 3. Cash Flow & Balance Sheet (FY2021–FY2025)

This section pulls a small set of cash flow and balance sheet items that feed directly into the DCF
we'll build later — the income statement alone isn't enough to get from profit to free cash flow, or
from enterprise value to equity value per share.

- **Depreciation and amortization (D&A)** — a non-cash expense already deducted in arriving at net
  income/EBIT. It gets added back when building unlevered free cash flow, since it doesn't consume cash.
- **Capital expenditure (Capex)** — cash spent on property, plant & equipment and on intangible assets.
  This *is* a real cash outflow (unlike D&A) and gets subtracted when building free cash flow — ASML is
  capital-intensive (building EUV/High-NA tool capacity), so this matters a lot.
- **Net cash flow from operating activities** — cash actually generated by the business, as reported in
  the cash flow statement. Useful as a cross-check against net income (earnings quality: is profit
  backed by real cash generation?).
- **Cash & cash equivalents (incl. short-term investments)** — non-operating liquid assets held at
  year-end. Needed later for the bridge from Enterprise Value to Equity Value (Equity Value =
  Enterprise Value − Debt + Cash).
- **Total debt (short-term + long-term borrowings)** and **Weighted average shares outstanding
  (basic/diluted)** — also needed for that same Enterprise-Value-to-Equity-Value-to-per-share bridge,
  and for capital structure weights in WACC.

**Source:** same five press releases already cited in Section 1, drawing this time on the linked
*"Financial statements US GAAP"* PDF exhibits' **"Summary US GAAP Consolidated Statements of Cash
Flows"** and **"Summary US GAAP Consolidated Balance Sheets"** tables (rather than the Statements of
Operations table used for Section 1).

**Flag — Total debt not fully disclosed:** these summary balance sheets show only a single aggregated
*"Current liabilities"* line (not broken into short-term debt vs. other current liabilities) and a
single *"Long-term debt"* line. There is no separately disclosed short-term/current-portion-of-debt
figure, so **"Total debt" cannot be reliably computed from the press release alone** and is left blank
below rather than guessed. The disclosed **Long-term debt** component is shown as a supplementary row
instead, clearly labelled as only part of total debt.

In [3]:
# Figures in EUR millions, except weighted average share counts which are in millions of shares.
# All figures are for the TWELVE MONTHS ENDED / AS AT December 31 of the stated year, taken from the
# "Summary US GAAP Consolidated Statements of Cash Flows" and "Summary US GAAP Consolidated Balance
# Sheets" tables of the linked "Financial statements US GAAP" PDF exhibit for each year's Q4 press
# release (same five press releases cited in Section 1).

cash_flow_bs_data = {
    "FY2021": {
        # Source: FY2021 press release (https://www.asml.com/en/news/press-releases/2022/q4-and-full-year-2021-financial-results)
        # -> "Financial statements US GAAP Q4 2021" PDF, p.3 (Cash Flows) & p.1 (Statements of
        # Operations, share counts), "Twelve months ended Dec 31, 2021"
        "D&A": 471.0,
        # Capex = Purchase of PP&E (900.7) + Purchase of intangible assets (39.6), same source p.3
        "Capex (PP&E + intangible assets)": 900.7 + 39.6,
        "Operating cash flow": 10845.8,
        # Balance sheet, p.2, "Dec 31, 2021": Cash and cash equivalents (6,951.8) + Short-term investments (638.5)
        "Cash & cash equivalents (incl. short-term investments)": 6951.8 + 638.5,
        "Total debt (ST + LT borrowings)": None,  # not disclosed - see markdown flag above
        "Long-term debt (disclosed component only)": 4075.0,  # balance sheet, p.2, "Dec 31, 2021"
        "Weighted avg shares - basic (millions)": 409.8,
        "Weighted avg shares - diluted (millions)": 410.4,
    },
    "FY2022": {
        # Source: FY2022 press release (https://www.asml.com/en/news/press-releases/2023/q4-2022-financial-results)
        # -> "Financial statements US GAAP Q4 2022" PDF, p.3 (Cash Flows) & p.1 (Statements of
        # Operations, share counts), "Twelve months ended Dec 31, 2022"
        "D&A": 583.6,
        # Capex = Purchase of PP&E (1,281.8) + Purchase of intangible assets (37.5), same source p.3
        "Capex (PP&E + intangible assets)": 1281.8 + 37.5,
        "Operating cash flow": 8486.8,
        # Balance sheet, p.2, "Dec 31, 2022": Cash and cash equivalents (7,268.3) + Short-term investments (107.7)
        "Cash & cash equivalents (incl. short-term investments)": 7268.3 + 107.7,
        "Total debt (ST + LT borrowings)": None,  # not disclosed - see markdown flag above
        "Long-term debt (disclosed component only)": 3514.2,  # balance sheet, p.2, "Dec 31, 2022"
        "Weighted avg shares - basic (millions)": 397.7,
        "Weighted avg shares - diluted (millions)": 398.1,
    },
    "FY2023": {
        # Source: FY2023 press release (https://www.asml.com/en/news/press-releases/2024/q4-2023-financial-results)
        # -> "Financial statements US GAAP Q4 2023" PDF, p.3 (Cash Flows) & p.1 (Statements of
        # Operations, share counts), "Twelve months ended Dec 31, 2023"
        "D&A": 739.8,
        # Capex = Purchase of PP&E (2,155.6) + Purchase of intangible assets (40.6), same source p.3
        "Capex (PP&E + intangible assets)": 2155.6 + 40.6,
        "Operating cash flow": 5443.4,
        # Balance sheet, p.2, "Dec 31, 2023": Cash and cash equivalents (7,004.7) + Short-term investments (5.4)
        "Cash & cash equivalents (incl. short-term investments)": 7004.7 + 5.4,
        "Total debt (ST + LT borrowings)": None,  # not disclosed - see markdown flag above
        "Long-term debt (disclosed component only)": 4631.5,  # balance sheet, p.2, "Dec 31, 2023"
        "Weighted avg shares - basic (millions)": 393.8,
        "Weighted avg shares - diluted (millions)": 394.1,
    },
    "FY2024": {
        # Source: FY2024 press release (https://www.asml.com/en/news/press-releases/2025/q4-2024-financial-results)
        # -> "Financial statements US GAAP Q4 2024" PDF, p.3 (Cash Flows) & p.1 (Statements of
        # Operations, share counts), "Twelve months ended Dec 31, 2024"
        "D&A": 918.6,
        # Capex = Purchase of PP&E (2,067.2) + Purchase of intangible assets (15.9), same source p.3
        "Capex (PP&E + intangible assets)": 2067.2 + 15.9,
        "Operating cash flow": 11166.2,
        # Balance sheet, p.2, "Dec 31, 2024": Cash and cash equivalents (12,735.9) + Short-term investments (5.4)
        "Cash & cash equivalents (incl. short-term investments)": 12735.9 + 5.4,
        "Total debt (ST + LT borrowings)": None,  # not disclosed - see markdown flag above
        "Long-term debt (disclosed component only)": 3677.3,  # balance sheet, p.2, "Dec 31, 2024"
        "Weighted avg shares - basic (millions)": 393.3,
        "Weighted avg shares - diluted (millions)": 393.6,
    },
    "FY2025": {
        # Source: FY2025 press release (https://www.asml.com/en/news/press-releases/2026/q4-2025-financial-results)
        # -> "Financial statements US GAAP Q4 2025" PDF, p.5 (Cash Flows) & p.3 (Statements of
        # Operations, share counts), "Twelve months ended Dec 31, 2025"
        "D&A": 1025.9,
        # Capex = Purchase of PP&E (1,573.6) + Purchase of intangible assets (57.6), same source p.5
        "Capex (PP&E + intangible assets)": 1573.6 + 57.6,
        "Operating cash flow": 12658.5,
        # Balance sheet, p.4, "Dec 31, 2025": Cash and cash equivalents (12,916.0) + Short-term investments (405.9)
        "Cash & cash equivalents (incl. short-term investments)": 12916.0 + 405.9,
        "Total debt (ST + LT borrowings)": None,  # not disclosed - see markdown flag above
        "Long-term debt (disclosed component only)": 2709.0,  # balance sheet, p.4, "Dec 31, 2025"
        "Weighted avg shares - basic (millions)": 388.5,
        "Weighted avg shares - diluted (millions)": 388.9,
    },
}

cash_flow_bs = pd.DataFrame(cash_flow_bs_data)
cash_flow_bs

,FY2021,FY2022,FY2023,FY2024,FY2025
D&A,471.0,583.6,739.8,918.6,1025.9
Capex (PP&E + intangible assets),940.3,1319.3,2196.2,2083.1,1631.2
Operating cash flow,10845.8,8486.8,5443.4,11166.2,12658.5
Cash & cash equivalents (incl. short-term investments),7590.3,7376.0,7010.1,12741.3,13321.9
Total debt (ST + LT borrowings),NaN,NaN,NaN,NaN,NaN
Long-term debt (disclosed component only),4075.0,3514.2,4631.5,3677.3,2709.0
Weighted avg shares - basic (millions),409.8,397.7,393.8,393.3,388.5
Weighted avg shares - diluted (millions),410.4,398.1,394.1,393.6,388.9


## 4. Qualitative Review (Strengths / Weaknesses / Opportunities / Threats)

This table captures how ASML's own management framed each year's results and outlook, in the
CEO/President's statement that opens each Q4 press release, plus (for risks specifically) the
"Forward Looking Statements" boilerplate section that every one of these press releases includes.

**Important caveat: this is management's own framing, not independent analysis.** A CEO statement is
written to tell a coherent, generally favorable story to investors — it will emphasize good news, and
it will rarely volunteer a self-critical "weakness." That's visible below: several years show "not
explicitly addressed in source" under Weaknesses, because the CEO commentary for that year genuinely
contains no self-identified shortfall — not because nothing ever went wrong, but because management
chose not to frame it that way in this document. This table is **one input among several** for
building a view of the business; it should be read alongside the historical financials above (which
are independently verifiable) and, eventually, alongside independent/analyst perspectives — not
treated as a substitute for either.

**Sourcing approach per cell:**
- **Strengths / Weaknesses / Opportunities** are drawn from the CEO/President's statement in each
  year's press release, quoted directly where possible.
- **Threats**: where the CEO statement itself names a specific risk, that's used and cited as
  "CEO statement." Where it doesn't (most years), the cell instead summarizes the press release's own
  **"Forward Looking Statements" section** — still part of the same press release, not the annual
  report — and is cited as such. The annual report was not needed for any cell below, since every one
  of these five press releases does contain its own Forward Looking Statements section.
- Where a category is genuinely not addressed by either of those sources for a given year, the cell
  says **"not explicitly addressed in source"** rather than inferring or inventing content.

In [4]:
qualitative_data = {
    "FY2021": {
        # Source: CEO (Peter Wennink) statement, FY2021 press release
        # (https://www.asml.com/en/news/press-releases/2022/q4-and-full-year-2021-financial-results)
        "Strengths": (
            "“For ASML, 2021 was a strong growth year in a dynamic environment” — net sales of "
            "€18.6bn (incl. €6.3bn from 42 EUV systems), net income of €5.9bn, 52.7% gross margin, "
            "and a 100% dividend increase to €5.50/share. (CEO statement)"
        ),
        "Weaknesses": (
            "The fire at ASML's Berlin site is flagged as a factor in 2022 guidance, and roughly €2bn of "
            "revenue is noted as shifting from Q1 2022 into later quarters due to fast shipments. (CEO statement)"
        ),
        "Opportunities": (
            "First order received for the next-gen TWINSCAN EXE:5200 (0.55NA/High-NA EUV) and a new NXT:870 "
            "KrF product planned for 2022; “we experience higher demand for our systems than our production "
            "capacity can accommodate.” (CEO statement)"
        ),
        "Threats": (
            "Press release's Forward Looking Statements section flags COVID-19 impact on operations/suppliers/"
            "financial markets, supply chain and logistics constraints, the Berlin fire's effect on production, "
            "and trade/import-export/national-security regulations. (Forward Looking Statements section, same press release)"
        ),
    },
    "FY2022": {
        # Source: CEO (Peter Wennink) statement, FY2022 press release
        # (https://www.asml.com/en/news/press-releases/2023/q4-2022-financial-results)
        "Strengths": (
            "“Total net sales for the year of €21.2 billion, gross margin of 50.5% and a record "
            "backlog at the end of 2022 of €40.4 billion”; Q4 gross margin of 51.5% beat guidance "
            "“due to additional upgrades and insurance settlement.” (CEO statement)"
        ),
        "Weaknesses": (
            "CEO cites “uncertainty in the market caused by inflation, rising interest rates, risk of "
            "recession and geopolitical developments related to export controls” as a near-term backdrop. "
            "(CEO statement)"
        ),
        "Opportunities": (
            "“ASML expects continued strong growth with a net sales increase of more than 25%” in "
            "2023, underpinned by long order lead times and “the strategic nature of lithography "
            "investments.” (CEO statement)"
        ),
        "Threats": (
            "CEO statement itself names inflation, rising interest rates, recession risk, and export-control-"
            "related geopolitical developments; the Forward Looking Statements section additionally cites "
            "supply chain constraints and the Russia-Ukraine conflict's effect on the global economy. "
            "(CEO statement + Forward Looking Statements section, same press release)"
        ),
    },
    "FY2023": {
        # Source: CEO (Peter Wennink) statement, FY2023 press release
        # (https://www.asml.com/en/news/press-releases/2024/q4-2023-financial-results)
        "Strengths": (
            "“ASML achieved another strong year in 2023 with 30% growth, ending with total net sales for "
            "the year of €27.6 billion”; shipped the first modules of the first High-NA EUV system "
            "(EXE:5000) to a customer. (CEO statement)"
        ),
        "Weaknesses": (
            "“The semiconductor industry continues to work through the bottom of the cycle” and "
            "customers remained “not certain about the shape of the semiconductor market recovery this "
            "year.” (CEO statement)"
        ),
        "Opportunities": (
            "Improving end-market inventory levels and litho tool utilization; strong Q4 order intake "
            "“clearly supports future demand”, with 2024 positioned as a preparation year for "
            "“significant growth that we expect for 2025.” (CEO statement)"
        ),
        "Threats": (
            "CEO commentary does not detail specific risks; the press release's Forward Looking Statements "
            "section flags export controls, geopolitical developments, and supply chain constraints. "
            "(Forward Looking Statements section, same press release)"
        ),
    },
    "FY2024": {
        # Source: CEO (Christophe Fouquet) statement, FY2024 press release
        # (https://www.asml.com/en/news/press-releases/2025/q4-2024-financial-results)
        "Strengths": (
            "“ASML achieved another record year, ending with total net sales for 2024 of €28.3 "
            "billion, and a gross margin of 51.3%”, with a record Q4 revenue of €9.3bn. (CEO statement)"
        ),
        "Weaknesses": "not explicitly addressed in source (CEO commentary is uniformly positive on FY2024 itself)",
        "Opportunities": (
            "“The growth in artificial intelligence is the key driver for growth in our industry”, "
            "which “has created a shift in the market dynamics.” (CEO statement)"
        ),
        "Threats": (
            "CEO notes the AI-driven shift “is not benefiting all of our customers equally, which creates "
            "both opportunities and risks”, reflected in a wide 2025 guidance range (€30–35bn); "
            "the Forward Looking Statements section separately lists geopolitical risk, export controls, and "
            "supply chain constraints. (CEO statement + Forward Looking Statements section, same press release)"
        ),
    },
    "FY2025": {
        # Source: CEO (Christophe Fouquet) statement, FY2025 press release
        # (https://www.asml.com/en/news/press-releases/2026/q4-2025-financial-results)
        "Strengths": (
            "“ASML reported another record year in 2025, with total net sales of €32.7 billion and a "
            "gross margin of 52.8%”; Q4 was “particularly strong” with record net sales of "
            "€9.7bn and record order intake. (CEO statement)"
        ),
        "Weaknesses": "not explicitly addressed in source (CEO commentary is uniformly positive on FY2025 itself)",
        "Opportunities": (
            "Customers gave “a notably more positive assessment of the medium-term market situation, "
            "primarily based on more robust expectations of the sustainability of AI-related demand”; CEO "
            "expects “2026 to be another growth year”, largely driven by EUV sales growth. (CEO statement)"
        ),
        "Threats": (
            "CEO statement itself names no explicit risks; the Forward Looking Statements section cites "
            "geopolitical risk, export controls, tariffs, and macroeconomic/geopolitical uncertainty as ongoing "
            "concerns. (Forward Looking Statements section, same press release)"
        ),
    },
}

qualitative_review = pd.DataFrame(qualitative_data).T[["Strengths", "Weaknesses", "Opportunities", "Threats"]]

with pd.option_context("display.max_colwidth", None):
    display(qualitative_review)

,Strengths,Weaknesses,Opportunities,Threats
FY2021,"“For ASML, 2021 was a strong growth year in a dynamic environment” — net sales of €18.6bn (incl. €6.3bn from 42 EUV systems), net income of €5.9bn, 52.7% gross margin, and a 100% dividend increase to €5.50/share. (CEO statement)","The fire at ASML's Berlin site is flagged as a factor in 2022 guidance, and roughly €2bn of revenue is noted as shifting from Q1 2022 into later quarters due to fast shipments. (CEO statement)",First order received for the next-gen TWINSCAN EXE:5200 (0.55NA/High-NA EUV) and a new NXT:870 KrF product planned for 2022; “we experience higher demand for our systems than our production capacity can accommodate.” (CEO statement),"Press release's Forward Looking Statements section flags COVID-19 impact on operations/suppliers/financial markets, supply chain and logistics constraints, the Berlin fire's effect on production, and trade/import-export/national-security regulations. (Forward Looking Statements section, same press release)"
FY2022,"“Total net sales for the year of €21.2 billion, gross margin of 50.5% and a record backlog at the end of 2022 of €40.4 billion”; Q4 gross margin of 51.5% beat guidance “due to additional upgrades and insurance settlement.” (CEO statement)","CEO cites “uncertainty in the market caused by inflation, rising interest rates, risk of recession and geopolitical developments related to export controls” as a near-term backdrop. (CEO statement)","“ASML expects continued strong growth with a net sales increase of more than 25%” in 2023, underpinned by long order lead times and “the strategic nature of lithography investments.” (CEO statement)","CEO statement itself names inflation, rising interest rates, recession risk, and export-control-related geopolitical developments; the Forward Looking Statements section additionally cites supply chain constraints and the Russia-Ukraine conflict's effect on the global economy. (CEO statement + Forward Looking Statements section, same press release)"
FY2023,"“ASML achieved another strong year in 2023 with 30% growth, ending with total net sales for the year of €27.6 billion”; shipped the first modules of the first High-NA EUV system (EXE:5000) to a customer. (CEO statement)",“The semiconductor industry continues to work through the bottom of the cycle” and customers remained “not certain about the shape of the semiconductor market recovery this year.” (CEO statement),"Improving end-market inventory levels and litho tool utilization; strong Q4 order intake “clearly supports future demand”, with 2024 positioned as a preparation year for “significant growth that we expect for 2025.” (CEO statement)","CEO commentary does not detail specific risks; the press release's Forward Looking Statements section flags export controls, geopolitical developments, and supply chain constraints. (Forward Looking Statements section, same press release)"
FY2024,"“ASML achieved another record year, ending with total net sales for 2024 of €28.3 billion, and a gross margin of 51.3%”, with a record Q4 revenue of €9.3bn. (CEO statement)",not explicitly addressed in source (CEO commentary is uniformly positive on FY2024 itself),"“The growth in artificial intelligence is the key driver for growth in our industry”, which “has created a shift in the market dynamics.” (CEO statement)","CEO notes the AI-driven shift “is not benefiting all of our customers equally, which creates both opportunities and risks”, reflected in a wide 2025 guidance range (€30–35bn); the Forward Looking Statements section separately lists geopolitical risk, export controls, and supply chain constraints. (CEO statement + Forward Looking Statements section, same press release)"
FY2025,"“ASML reported another record year in 2025, with total net sales of €32.7 billion and a gross margin of 52.8%”; Q4 was “particularly strong” with record net sales of €9.7bn and record order intake. (CEO statement)",not explicitly addressed in source (CEO commentary is

## 5. Forecast Assumptions (FY2026–FY2030)

**This section is fundamentally different from Sections 1–4 above.** Everything before this point was
either a sourced, cited fact (Sections 1 and 3), a formula computed from sourced facts (Section 2), or
management's own quoted words (Section 4). Nothing below this line is a fact — it's a set of **forward
assumptions**: judgment calls about how ASML's revenue, margins, tax rate, capex, D&A, and working
capital might evolve over the next five years, which will later feed into the actual DCF forecast.

Each assumption is anchored to *something* — a specific external source, a historical average, or a
reasoned interpolation between two sourced anchor points — but "anchored to" is not the same as
"derived from." These are estimates, not facts, and they are **meant to be reviewed, argued with, and
edited** before they feed into any forecast. The comments in the code cell below, and the markdown
synopsis immediately beneath it, explain the specific reasoning and source behind each number.

A few general notes on the approach:
- **Revenue growth % and Gross margin % are now year-by-year, each pinned to a specific external
  source** (ASML's own raised FY2026 guidance, a third-party SEMI industry equipment-spending
  forecast, and ASML's own November 2024 Investor Day 2030 scenario) rather than a smooth
  mathematical fade — see the synopsis below the code cell for the source behind each individual
  year. One number (FY2030 revenue growth) is flagged as an unsourced judgment call rather than
  pinned to an external figure.
- **R&D%, SG&A% cost ratios** are anchored to the range and average already visible in Section 2's
  margin table, generally held close to recent actuals with only modest, explicitly-reasoned drift.
- **The effective tax rate assumption is calculated, not chosen** — it's pulled directly from the
  5-year average of the "Effective tax rate %" row added to Section 1 above, so it updates automatically
  if the historical inputs ever change.
- **Capex% and D&A%** are anchored to the ranges implied by Section 3's Capex and D&A rows (as a % of
  the same year's net sales from Section 1).
- **Change in net working capital (ΔNWC) % of net sales is the weakest-anchored assumption here** — this
  notebook has not pulled in enough balance sheet detail (inventory, receivables, payables, contract
  assets/liabilities) to calculate a historical NWC figure, so this one is a standalone placeholder
  judgment call, flagged clearly as such rather than dressed up as more rigorous than it is.

In [5]:
# Historical anchor points pulled from earlier sections (not re-typed by hand, so these stay correct
# if the historical inputs above are ever revised).
historical_capex_pct = cash_flow_bs.loc["Capex (PP&E + intangible assets)"] / historicals.loc["Net sales"] * 100
historical_da_pct = cash_flow_bs.loc["D&A"] / historicals.loc["Net sales"] * 100
historical_tax_rate_avg = historicals.loc["Effective tax rate %"].mean()

print("Historical anchor points (FY2021-FY2025):")
print(f"  Capex % of net sales: {historical_capex_pct.min():.1f}% - {historical_capex_pct.max():.1f}% "
      f"(avg {historical_capex_pct.mean():.1f}%)")
print(f"  D&A % of net sales:   {historical_da_pct.min():.1f}% - {historical_da_pct.max():.1f}% "
      f"(avg {historical_da_pct.mean():.1f}%)")
print(f"  Effective tax rate:   avg {historical_tax_rate_avg:.1f}% <- this literally sets the assumption below")

# ==============================================================================================
#  FORECAST ASSUMPTIONS - FY2026-FY2030
#  ------------------------------------
#  THESE ARE JUDGMENT CALLS, NOT SOURCED FACTS. Review, argue with, and edit every number below
#  before it feeds into any forecast. See the markdown cell above for the general reasoning, and
#  the markdown synopsis below this cell for the specific source behind each Revenue growth % and
#  Gross margin % figure.
# ==============================================================================================

assumptions = {
    "FY2026": {
        # Revenue growth: 35%. ASML's own raised FY2026 guidance of EUR 43-45bn (midpoint EUR 44bn)
        # vs. FY2025 actual net sales of EUR 32,667.3m (Section 1). Source: BigGo Finance, reporting
        # on ASML's Q2 2026 earnings release, July 2026
        # (https://finance.biggo.com/news/69d3fa9d-7754-437a-aafb-4a8b34b2a5ac).
        "Revenue growth %": 35.0,
        # Gross margin: 53.5%. Reasoned interpolation between FY2025 actual (52.8%, Section 2) and
        # ASML's reaffirmed FY2030 gross margin scenario of 56%-60% (same Nov 2024 Investor Day
        # source cited below) - not an individually sourced figure for FY2026 specifically.
        "Gross margin %": 53.5,
        "R&D % of net sales": 14.2,
        "SG&A % of net sales": 4.0,
        "Effective tax rate %": round(historical_tax_rate_avg, 1),
        "Capex % of net sales": 7.0,
        "D&A % of net sales": 3.2,
        "Change in NWC % of net sales": 2.0,
    },
    "FY2027": {
        # Revenue growth: 14%. SEMI's global 300mm fab equipment spending forecast projects +14%
        # industry-wide spending growth for 2027 (to $151bn). Source: SEMI press release, "SEMI
        # Projects Double-Digit Growth in Global 300mm Fab Equipment Spending for 2026 and 2027"
        # (https://www.semi.org/en/semi-press-release/semi-projects-double-digit-growth-in-global-300mm-fab-equipment-spending-for-2026-and-2027).
        # Used as a proxy for ASML's own growth, since ASML sells directly into this capex cycle.
        "Revenue growth %": 14.0,
        "Gross margin %": 54.5,          # interpolation toward the FY2030 target range - see FY2026 note
        "R&D % of net sales": 14.0,
        "SG&A % of net sales": 3.9,
        "Effective tax rate %": round(historical_tax_rate_avg, 1),
        "Capex % of net sales": 6.5,
        "D&A % of net sales": 3.2,
        "Change in NWC % of net sales": 2.0,
    },
    "FY2028": {
        # Revenue growth: 3%. Same SEMI forecast projects industry-wide 300mm fab equipment spending
        # decelerating sharply to +3% in 2028 (to $155bn) before reaccelerating in 2029 - used here as
        # a proxy for ASML, same source as FY2027 above.
        "Revenue growth %": 3.0,
        "Gross margin %": 54.0,          # dips, tracking the FY2028 growth deceleration - judgment call
        "R&D % of net sales": 13.8,
        "SG&A % of net sales": 3.9,
        "Effective tax rate %": round(historical_tax_rate_avg, 1),
        "Capex % of net sales": 6.0,
        "D&A % of net sales": 3.1,
        "Change in NWC % of net sales": 2.0,
    },
    "FY2029": {
        # Revenue growth: 11%. Same SEMI forecast projects a reacceleration to +11% industry-wide
        # spending growth in 2029 (to $172bn) - same source as FY2027/FY2028 above.
        "Revenue growth %": 11.0,
        "Gross margin %": 56.0,          # interpolation, now at the bottom of the FY2030 target range
        "R&D % of net sales": 13.7,
        "SG&A % of net sales": 3.9,
        "Effective tax rate %": round(historical_tax_rate_avg, 1),
        "Capex % of net sales": 5.5,
        "D&A % of net sales": 3.1,
        "Change in NWC % of net sales": 2.0,
    },
    "FY2030": {
        # Revenue growth: 8% - FLAGGED AS THE ONE UNSOURCED JUDGMENT CALL IN THIS ROW. Not pinned to
        # an external figure; chosen so that FY2030 implied revenue lands modestly above the top of
        # ASML's own reaffirmed 2030 revenue scenario of EUR 44bn-60bn (Nov 2024 Investor Day press
        # release, https://www.asml.com/en/news/press-releases/2024/asml-investor-day-2024), since the
        # FY2026 guidance jump alone (EUR ~44.1bn implied) already sits near the bottom of that range,
        # so holding FY2027-FY2029 SEMI-based growth on top of it would otherwise imply 2030 revenue
        # far above ASML's own scenario ceiling without a specific reason to override it. See the
        # implied revenue calculation below the assumptions table for the actual EUR result.
        "Revenue growth %": 8.0,
        # Gross margin: 58%, the midpoint of ASML's reaffirmed FY2030 gross margin scenario of
        # 56%-60% (same Nov 2024 Investor Day source above) - this is the one row-5 figure that IS
        # directly sourced, not interpolated.
        "Gross margin %": 58.0,
        "R&D % of net sales": 13.6,
        "SG&A % of net sales": 3.9,
        "Effective tax rate %": round(historical_tax_rate_avg, 1),
        "Capex % of net sales": 5.0,
        "D&A % of net sales": 3.0,
        "Change in NWC % of net sales": 2.0,
    },
}

forecast_assumptions = pd.DataFrame(assumptions)
forecast_assumptions_display = forecast_assumptions.map(lambda v: f"{v:.1f}%")
display(forecast_assumptions_display)

# Implied revenue and gross profit in EUR millions, compounding the Revenue growth % assumptions
# forward from the FY2025 actual net sales figure in Section 1 (not re-typed by hand).
implied = pd.DataFrame(index=["Implied net sales", "Implied gross profit"], columns=forecast_assumptions.columns, dtype=float)
prior_sales = historicals.loc["Net sales", "FY2025"]
for year in forecast_assumptions.columns:
    sales = prior_sales * (1 + forecast_assumptions.loc["Revenue growth %", year] / 100)
    implied.loc["Implied net sales", year] = sales
    implied.loc["Implied gross profit", year] = sales * forecast_assumptions.loc["Gross margin %", year] / 100
    prior_sales = sales

implied_display = implied.map(lambda v: f"€{v:,.0f}m")
display(implied_display)

Historical anchor points (FY2021-FY2025):
  Capex % of net sales: 5.0% - 8.0% (avg 6.3%)
  D&A % of net sales:   2.5% - 3.3% (avg 2.9%)
  Effective tax rate:   avg 16.5% <- this literally sets the assumption below


,FY2026,FY2027,FY2028,FY2029,FY2030
Revenue growth %,35.0%,14.0%,3.0%,11.0%,8.0%
Gross margin %,53.5%,54.5%,54.0%,56.0%,58.0%
R&D % of net sales,14.2%,14.0%,13.8%,13.7%,13.6%
SG&A % of net sales,4.0%,3.9%,3.9%,3.9%,3.9%
Effective tax rate %,16.5%,16.5%,16.5%,16.5%,16.5%
Capex % of net sales,7.0%,6.5%,6.0%,5.5%,5.0%
D&A % of net sales,3.2%,3.2%,3.1%,3.1%,3.0%
Change in NWC % of net sales,2.0%,2.0%,2.0%,2.0%,2.0%


,FY2026,FY2027,FY2028,FY2029,FY2030
Implied net sales,"€44,101m","€50,275m","€51,783m","€57,479m","€62,078m"
Implied gross profit,"€23,594m","€27,400m","€27,963m","€32,188m","€36,005m"


### Synopsis: Revenue growth % and Gross margin % sources

**Revenue growth %**
- **FY2026 — 35%.** ASML's own raised FY2026 guidance of €43–45bn (midpoint €44bn) vs. FY2025 actual
  net sales of €32,667.3m (Section 1). Source: [BigGo Finance, reporting ASML's Q2 2026 earnings
  release](https://finance.biggo.com/news/69d3fa9d-7754-437a-aafb-4a8b34b2a5ac), July 2026.
- **FY2027 — 14%.** SEMI's global 300mm fab equipment spending forecast projects industry-wide
  spending +14% in 2027, used as a proxy for ASML's own growth. Source:
  [SEMI press release](https://www.semi.org/en/semi-press-release/semi-projects-double-digit-growth-in-global-300mm-fab-equipment-spending-for-2026-and-2027).
- **FY2028 — 3%.** Same SEMI forecast, which projects a sharp industry-wide deceleration to +3% in
  2028 before reaccelerating — same source as FY2027.
- **FY2029 — 11%.** Same SEMI forecast's projected reacceleration to +11% in 2029 — same source as
  FY2027/FY2028.
- **FY2030 — 8%. Flagged: this one is an unsourced judgment call**, not pinned to an external figure.
  It's chosen so implied FY2030 revenue lands modestly above the top of ASML's own reaffirmed 2030
  revenue scenario (below), since FY2026 guidance alone already implies revenue near the bottom of
  that range.

**Gross margin %**
- **FY2030 — 58%.** The midpoint of ASML's own reaffirmed 2030 gross margin scenario, "approximately
  56% and 60%," announced at its November 2024 Investor Day. Source:
  [ASML Investor Day 2024 press release](https://www.asml.com/en/news/press-releases/2024/asml-investor-day-2024).
  This is the one gross-margin figure in this table that's directly sourced rather than interpolated.
- **FY2026–FY2029 — 53.5% / 54.5% / 54.0% / 56.0%.** A reasoned interpolation between FY2025 actual
  (52.8%, Section 2) and the FY2030 target above — **flagged as interpolated, not individually
  sourced for each of these four years.** FY2028 dips slightly to track that year's SEMI-implied
  growth deceleration, rather than rising smoothly.

The implied revenue and gross profit table above (printed by the same code cell, right after the
percentage assumptions table) simply compounds the Revenue growth % row forward from the FY2025
actual and multiplies by the Gross margin % row — it isn't a separate assumption, just these two
rows expressed in euros instead of percent.

## 6. Forecast Financials & Unlevered FCF (FY2026–FY2030)

**This section is purely mechanical — every judgment call already happened in Section 5.** Nothing
here is hardcoded; every cell below is a formula that reads the percentage assumptions from Section 5
(and the FY2025 actual net sales from Section 1) and mechanically works through to unlevered free
cash flow (UFCF). If you want to change the forecast, go edit Section 5 — this section will update
automatically.

The build, step by step:
1. **Net sales** and **Gross profit** — compounding the Revenue growth % assumption forward from the
   FY2025 actual, then applying the Gross margin % assumption (same calculation Section 5's "Implied"
   table already did — reused here rather than recomputed).
2. **R&D expenses** and **SG&A expenses** — each year's Net sales × that year's R&D%/SG&A% assumption.
3. **EBIT** = Gross profit − R&D expenses − SG&A expenses — operating profit before interest and tax.
4. **Tax on EBIT** = EBIT × Effective tax rate % — the cash tax bill on operating profit, ignoring
   financing effects (this is the "unlevered" part: we tax EBIT directly, not profit after interest).
5. **NOPAT** (Net Operating Profit After Tax) = EBIT − Tax on EBIT — what operating profit would be if
   the company had no debt.
6. **D&A** is added back — it was already subtracted in arriving at EBIT/NOPAT, but it's a non-cash
   accounting charge, so it needs adding back to get to a *cash* flow figure.
7. **Capex** is subtracted — the real cash spent on property, plant, equipment and intangible assets,
   which never shows up in the income statement at all but very much shows up in the cash flow.
8. **Change in net working capital (ΔNWC)** is subtracted — cash tied up funding growth (more
   inventory, more receivables, etc. as the business gets bigger), separate from capex or D&A.
9. **Unlevered Free Cash Flow** = NOPAT + D&A − Capex − ΔNWC — the cash the business generates before
   any financing decisions (debt/equity), which is exactly the cash flow a DCF discounts at WACC to
   get Enterprise Value.

In [6]:
forecast = pd.DataFrame(index=[
    "Net sales",
    "Gross profit",
    "R&D expenses",
    "SG&A expenses",
    "EBIT",
    "Tax on EBIT",
    "NOPAT",
    "D&A",
    "Capex",
    "Change in NWC",
    "Unlevered Free Cash Flow",
], columns=forecast_assumptions.columns, dtype=float)

# Net sales and Gross profit: reuse Section 5's "implied" figures directly (same numbers, not
# recomputed) rather than duplicating the compounding logic here.
forecast.loc["Net sales"] = implied.loc["Implied net sales"]
forecast.loc["Gross profit"] = implied.loc["Implied gross profit"]

# R&D and SG&A: this year's Net sales x that year's R&D%/SG&A% assumption from Section 5.
forecast.loc["R&D expenses"] = forecast.loc["Net sales"] * forecast_assumptions.loc["R&D % of net sales"] / 100
forecast.loc["SG&A expenses"] = forecast.loc["Net sales"] * forecast_assumptions.loc["SG&A % of net sales"] / 100

forecast.loc["EBIT"] = forecast.loc["Gross profit"] - forecast.loc["R&D expenses"] - forecast.loc["SG&A expenses"]

# Tax on EBIT and NOPAT: apply Section 5's effective tax rate assumption directly to EBIT (the
# "unlevered" step - tax is applied to operating profit, not profit after interest).
tax_rate = forecast_assumptions.loc["Effective tax rate %"] / 100
forecast.loc["Tax on EBIT"] = forecast.loc["EBIT"] * tax_rate
forecast.loc["NOPAT"] = forecast.loc["EBIT"] * (1 - tax_rate)

# D&A, Capex, Change in NWC: this year's Net sales x that year's %-of-net-sales assumption from
# Section 5. All three are stored as positive amounts here (D&A added back, Capex and NWC subtracted
# below), matching the formula as specified rather than the negative-expense convention used in
# Section 1's historical table.
forecast.loc["D&A"] = forecast.loc["Net sales"] * forecast_assumptions.loc["D&A % of net sales"] / 100
forecast.loc["Capex"] = forecast.loc["Net sales"] * forecast_assumptions.loc["Capex % of net sales"] / 100
forecast.loc["Change in NWC"] = forecast.loc["Net sales"] * forecast_assumptions.loc["Change in NWC % of net sales"] / 100

forecast.loc["Unlevered Free Cash Flow"] = (
    forecast.loc["NOPAT"] + forecast.loc["D&A"] - forecast.loc["Capex"] - forecast.loc["Change in NWC"]
)

# Sanity check: EBIT should split exactly into Tax on EBIT + NOPAT for every year.
assert (forecast.loc["Tax on EBIT"] + forecast.loc["NOPAT"] - forecast.loc["EBIT"]).abs().max() < 1e-6, \
    "EBIT does not split into Tax on EBIT + NOPAT"

forecast_display = forecast.map(lambda v: f"€{v:,.0f}m")
display(forecast_display)

,FY2026,FY2027,FY2028,FY2029,FY2030
Net sales,"€44,101m","€50,275m","€51,783m","€57,479m","€62,078m"
Gross profit,"€23,594m","€27,400m","€27,963m","€32,188m","€36,005m"
R&D expenses,"€6,262m","€7,038m","€7,146m","€7,875m","€8,443m"
SG&A expenses,"€1,764m","€1,961m","€2,020m","€2,242m","€2,421m"
EBIT,"€15,568m","€18,401m","€18,797m","€22,072m","€25,141m"
Tax on EBIT,"€2,569m","€3,036m","€3,102m","€3,642m","€4,148m"
NOPAT,"€12,999m","€15,365m","€15,696m","€18,430m","€20,993m"
D&A,"€1,411m","€1,609m","€1,605m","€1,782m","€1,862m"
Capex,"€3,087m","€3,268m","€3,107m","€3,161m","€3,104m"
Change in NWC,€882m,"€1,005m","€1,036m","€1,150m","€1,242m"


## 7. WACC (Discount Rate)

The Weighted Average Cost of Capital (WACC) is the rate we'll use to discount the unlevered free
cash flows from Section 6 back to present value — it represents the blended return that both equity
holders and debt holders require to fund ASML, weighted by how much of the company is actually
funded by each.

**Every input below is hardcoded and individually sourced** (this section behaves like Sections 1 and
3 — sourced facts — rather than Section 5's judgment-call assumptions), with one explicit exception
flagged below. The four WACC formulas themselves (cost of equity, after-tax cost of debt, capital
weights, WACC) are all computed, not hardcoded.

**Key takeaway — capital structure is overwhelmingly equity-funded.** ASML's FY2025 long-term debt
(€2,709.0m, Section 3) is under 1% of its market capitalization (~€533bn). With debt weight that
small, WACC ends up very close to the cost of equity alone — the after-tax cost of debt barely moves
the blended number, no matter how it's estimated.

**The one estimated (not directly sourced) input is the credit spread used to build the pre-tax cost
of debt.** ASML's actual traded bond yields would be the correct input, but they sit behind paywalls
on every source checked; a ~0.70% credit spread over the risk-free rate (consistent with ASML's
Moody's A2 rating) is used instead as a reasoned estimate, clearly flagged in the code below. Given
how small debt's weight is in the blend, even a meaningfully wrong estimate here would barely move
the final WACC.

In [7]:
# ==============================================================================================
#  WACC INPUTS - hardcoded, individually sourced (one estimate flagged explicitly below)
# ==============================================================================================

# Risk-free rate: German 10-year Bund yield, Sept 14 2026.
# Source: https://www.thetrading.tools/germany-10-year-yield ("Germany 10Y stands at 3.55%")
risk_free_rate = 3.55 / 100

# Beta: ASML's (levered) beta.
# Source: https://stockanalysis.com/stocks/asml/, Sept 14 2026 ("Beta: 1.36")
beta = 1.36

# Equity risk premium: Damodaran's Netherlands / mature-market total equity risk premium estimate.
# Source: https://pages.stern.nyu.edu/~adamodar/New_Home_Page/datafile/ctryprem.html
# (Netherlands row, 4.23%; page last updated January 2026)
equity_risk_premium = 4.23 / 100

# Market capitalization: built from two sourced primitives rather than hardcoding the converted
# EUR figure directly, so the conversion itself is a visible formula, not a re-typed number.
#   - Market cap in USD: https://stockanalysis.com/stocks/asml/, Sept 14 2026 ("Market Cap: 615.10B")
#   - EUR/USD exchange rate: https://tradingeconomics.com/euro-area/currency, Sept 14 2026 (~1.1537)
market_cap_usd_bn = 615.10
eur_usd_rate = 1.1537
market_cap_eur_m = market_cap_usd_bn * 1000 / eur_usd_rate  # -> ~EUR 533,100m

# Long-term debt: NOT re-hardcoded here - pulled directly from Section 3's already-sourced FY2025
# figure (cash_flow_bs), so this stays correct if Section 3 is ever revised.
long_term_debt_eur_m = cash_flow_bs.loc["Long-term debt (disclosed component only)", "FY2025"]

# Pre-tax cost of debt: built as Risk-free rate + an assumed credit spread, NOT independently
# sourced - ASML's actual traded bond yields were behind paywalls on every source checked.
# FLAGGED AS AN ESTIMATE: ~0.70% credit spread assumed, consistent with ASML's Moody's A2 rating.
credit_spread_estimate = 0.70 / 100  # ESTIMATE, not sourced - see markdown flag above
pre_tax_cost_of_debt = risk_free_rate + credit_spread_estimate

# Effective tax rate: NOT re-hardcoded - reused directly from Section 5's assumption (itself
# anchored to the Section 1 historical average), which is flat across the forecast so any column
# gives the same value.
tax_rate_for_wacc = forecast_assumptions.loc["Effective tax rate %", "FY2026"] / 100

# ==============================================================================================
#  WACC FORMULAS - all computed, nothing hardcoded below this line
# ==============================================================================================

cost_of_equity = risk_free_rate + beta * equity_risk_premium
after_tax_cost_of_debt = pre_tax_cost_of_debt * (1 - tax_rate_for_wacc)

total_capital = market_cap_eur_m + long_term_debt_eur_m
equity_weight = market_cap_eur_m / total_capital
debt_weight = long_term_debt_eur_m / total_capital

wacc = equity_weight * cost_of_equity + debt_weight * after_tax_cost_of_debt

wacc_inputs = pd.Series({
    "Risk-free rate": f"{risk_free_rate:.2%}",
    "Beta": f"{beta:.2f}",
    "Equity risk premium": f"{equity_risk_premium:.2%}",
    "Market capitalization (EUR m)": f"€{market_cap_eur_m:,.0f}m",
    "Long-term debt (EUR m, FY2025)": f"€{long_term_debt_eur_m:,.1f}m",
    "Pre-tax cost of debt (ESTIMATE)": f"{pre_tax_cost_of_debt:.2%}",
    "Effective tax rate": f"{tax_rate_for_wacc:.1%}",
})
print("WACC inputs:")
display(wacc_inputs)

wacc_outputs = pd.Series({
    "Cost of equity (CAPM)": f"{cost_of_equity:.2%}",
    "After-tax cost of debt": f"{after_tax_cost_of_debt:.2%}",
    "Equity weight": f"{equity_weight:.2%}",
    "Debt weight": f"{debt_weight:.2%}",
    "WACC": f"{wacc:.2%}",
})
print("\nWACC build:")
display(wacc_outputs)

print(f"\nWACC = {wacc:.2%} (vs. cost of equity alone = {cost_of_equity:.2%}) "
      f"- nearly identical, since debt weight is only {debt_weight:.2%} of total capital.")

WACC inputs:


Risk-free rate                         3.55%
Beta                                    1.36
Equity risk premium                    4.23%
Market capitalization (EUR m)      €533,154m
Long-term debt (EUR m, FY2025)     €2,709.0m
Pre-tax cost of debt (ESTIMATE)        4.25%
Effective tax rate                     16.5%
dtype: str


WACC build:


Cost of equity (CAPM)      9.30%
After-tax cost of debt     3.55%
Equity weight             99.49%
Debt weight                0.51%
WACC                       9.27%
dtype: str


WACC = 9.27% (vs. cost of equity alone = 9.30%) - nearly identical, since debt weight is only 0.51% of total capital.


## 8. Terminal Value (Gordon Growth vs Exit Multiple)

The 5-year explicit forecast (Sections 5–6) can't run forever — at some point we need a **terminal
value** that stands in for all the cash flows ASML generates beyond FY2030. This section builds that
terminal value two different ways and cross-checks them against each other, per the project's
approach of always building both methods rather than relying on one.

- **Gordon Growth (perpetuity growth) method** assumes FY2030's unlevered FCF grows forever at a
  constant, modest rate, discounted at WACC. It's a *cash-flow-based* method.
- **Exit Multiple method** assumes ASML gets sold, at the end of FY2030, for a multiple of its EBITDA
  that a buyer would pay today for a similar business. It's a *market-based* method, anchored to how
  the market actually prices comparable(-to-ASML-today) companies rather than a growth assumption.

**Expect these two methods to disagree, and expect the gap to be large here.** A textbook "mature,
stable" company would see the two converge, because a market multiple on a mature business is itself
just a compressed way of pricing in a modest perpetual growth rate — similar to what Gordon Growth
assumes directly. ASML's 10-year median EV/EBITDA multiple (used deliberately here, instead of
today's elevated ~44x, specifically to avoid extrapolating a possibly-temporary AI-driven re-rating)
still implies a *much* higher perpetual growth rate than the ECB's 2% medium-term inflation target
used for Gordon Growth — because the market has historically been pricing ASML as a structural
growth business (EUV/High-NA monopoly, secular semiconductor capex growth), not a mature/stable one.
**A large gap here is the expected, correct result of that difference in method — not a bug to
reconcile away.** The code cell below explicitly back-solves the perpetual growth rate the exit
multiple is implying, so that gap is visible as a number, not just a vibe.

In [8]:
# ==============================================================================================
#  TERMINAL VALUE INPUTS - hardcoded, individually sourced
# ==============================================================================================

# Terminal growth rate: ECB's medium-term inflation target.
# Source: https://www.ecb.europa.eu/mopo/strategy/pricestab/html/index.en.html
# ("the ECB's Governing Council ... considers that price stability is best maintained by aiming
# for 2% inflation over the medium term")
terminal_growth_rate = 2.0 / 100

# Exit EV/EBITDA multiple: ASML's 10-year historical median (deliberately used instead of the
# current ~44x multiple, to avoid extrapolating today's possibly-elevated, AI-driven re-rating
# into a permanent assumption).
# Source: https://www.gurufocus.com/term/enterprise-value-to-ebitda/ASML, Sept 15 2026
# (page states ASML's current EV/EBITDA of 44.00 is "51% Above Median", with 10-year median 29.06 -
# 44.00 / 1.51 = 29.14, consistent with the stated 29.06 median within rounding)
exit_ev_ebitda_multiple = 29.06

# ==============================================================================================
#  TERMINAL VALUE FORMULAS - all computed from Section 6 (forecast) and Section 7 (wacc)
# ==============================================================================================

fy2030_ufcf = forecast.loc["Unlevered Free Cash Flow", "FY2030"]
fy2030_ebitda = forecast.loc["EBIT", "FY2030"] + forecast.loc["D&A", "FY2030"]

# Gordon Growth (perpetuity growth) terminal value, as of end of FY2030.
gordon_growth_tv = fy2030_ufcf * (1 + terminal_growth_rate) / (wacc - terminal_growth_rate)

# Exit Multiple terminal value, as of end of FY2030.
exit_multiple_tv = fy2030_ebitda * exit_ev_ebitda_multiple

# Cross-check: back-solve the Gordon Growth formula for the perpetual growth rate that would be
# needed to justify the Exit Multiple terminal value, given the same FY2030 UFCF and WACC.
#   TV = FCF*(1+g) / (WACC-g)  =>  g = (TV*WACC - FCF) / (TV + FCF)
implied_growth_from_exit_multiple = (
    (exit_multiple_tv * wacc - fy2030_ufcf) / (exit_multiple_tv + fy2030_ufcf)
)

terminal_value_summary = pd.Series({
    "FY2030 Unlevered FCF": f"€{fy2030_ufcf:,.0f}m",
    "FY2030 EBITDA (EBIT + D&A)": f"€{fy2030_ebitda:,.0f}m",
    "WACC (Section 7)": f"{wacc:.2%}",
    "Terminal growth rate (Gordon Growth)": f"{terminal_growth_rate:.2%}",
    "Gordon Growth terminal value": f"€{gordon_growth_tv:,.0f}m",
    "Exit EV/EBITDA multiple": f"{exit_ev_ebitda_multiple:.2f}x",
    "Exit Multiple terminal value": f"€{exit_multiple_tv:,.0f}m",
    "Implied perpetual growth rate from Exit Multiple": f"{implied_growth_from_exit_multiple:.2%}",
})
display(terminal_value_summary)

gap_x = exit_multiple_tv / gordon_growth_tv
print(f"\nExit Multiple terminal value is {gap_x:.1f}x the Gordon Growth terminal value.")
print(f"The exit multiple is implicitly pricing in ~{implied_growth_from_exit_multiple:.1%} perpetual "
      f"growth, vs. the {terminal_growth_rate:.1%} used directly in the Gordon Growth method.")

FY2030 Unlevered FCF                                 €18,510m
FY2030 EBITDA (EBIT + D&A)                           €27,004m
WACC (Section 7)                                        9.27%
Terminal growth rate (Gordon Growth)                    2.00%
Gordon Growth terminal value                        €259,568m
Exit EV/EBITDA multiple                                29.06x
Exit Multiple terminal value                        €784,731m
Implied perpetual growth rate from Exit Multiple        6.76%
dtype: str


Exit Multiple terminal value is 3.0x the Gordon Growth terminal value.
The exit multiple is implicitly pricing in ~6.8% perpetual growth, vs. the 2.0% used directly in the Gordon Growth method.


## 9. Valuation: EV to Equity Value to Per-Share

This is the final bridge: discount the explicit FY2026–FY2030 unlevered FCFs (Section 6) and both
Section 8 terminal values back to present value at WACC (Section 7), sum to Enterprise Value, then
walk EV down to an equity value and a per-share figure — **twice**, once per terminal value method.

**Why a range, not one number.** Gordon Growth and Exit Multiple aren't two attempts at the same
answer where one is "right" — they encode two different beliefs about what ASML becomes after 2030:
- The **Gordon Growth price is what ASML is worth if it settles into a mature, roughly
  GDP/inflation-rate-growing business** after 2030 — the 2% ECB inflation-anchored terminal growth
  rate from Section 8 treats FY2030 as basically "steady state."
- The **Exit Multiple price is what ASML is worth if the market keeps pricing it as a durable
  secular grower** — Section 8 showed this method's 10-year-median multiple is implicitly pricing in
  ~6.8% perpetual growth, well above GDP/inflation, consistent with ASML's structural EUV/High-NA
  position rather than a maturing business.

Both are defensible; which one is closer to reality depends on a view of ASML's competitive position
in the 2030s that this notebook hasn't tried to independently adjudicate (that's exactly the kind of
qualitative judgment Section 4 flagged as "one input among several," not something a spreadsheet
settles). **The honest output of this model is therefore the range between the two, not a single
blended midpoint** — collapsing it into one number would hide that those two beliefs disagree by a
lot, which is real information, not noise to average away.

**Net debt sign check:** ASML holds far more cash (€13,321.9m, Section 3, FY2025) than long-term debt
(€2,709.0m, Section 7), so it's in a **net cash position** — net debt is *negative*. The bridge below
uses the standard formula Equity Value = EV − Net Debt with Net Debt = Debt − Cash; because Net Debt
comes out negative here, subtracting it *adds* the net cash to EV automatically — the formula doesn't
need special-casing, but the code below prints the net debt figure explicitly so that sign is visible
rather than just implicit in the arithmetic.

In [9]:
# ==============================================================================================
#  DISCOUNT EXPLICIT-PERIOD FCF AND BOTH TERMINAL VALUES TO PRESENT VALUE
# ==============================================================================================

forecast_years = forecast.columns.tolist()  # FY2026..FY2030
discount_periods = pd.Series(range(1, len(forecast_years) + 1), index=forecast_years)  # year 1..5
discount_factors = 1 / (1 + wacc) ** discount_periods

pv_fcf = forecast.loc["Unlevered Free Cash Flow"] * discount_factors

pv_build = pd.DataFrame({
    "Unlevered FCF": forecast.loc["Unlevered Free Cash Flow"],
    "Discount period": discount_periods,
    "Discount factor": discount_factors,
    "PV of FCF": pv_fcf,
})
pv_build_display = pv_build.copy()
pv_build_display["Unlevered FCF"] = pv_build_display["Unlevered FCF"].map(lambda v: f"€{v:,.0f}m")
pv_build_display["Discount factor"] = pv_build_display["Discount factor"].map(lambda v: f"{v:.4f}")
pv_build_display["PV of FCF"] = pv_build_display["PV of FCF"].map(lambda v: f"€{v:,.0f}m")
print("Present value of each explicit-period FCF (Section 6 FCF x discount factor at WACC):")
display(pv_build_display.T)

sum_pv_fcf = pv_fcf.sum()

# Both terminal values (Section 8) are "as of end of FY2030" - same discount factor as the FY2030
# FCF (year 5), per the brief.
year5_discount_factor = discount_factors["FY2030"]
pv_gordon_growth_tv = gordon_growth_tv * year5_discount_factor
pv_exit_multiple_tv = exit_multiple_tv * year5_discount_factor

ev_gordon_growth = sum_pv_fcf + pv_gordon_growth_tv
ev_exit_multiple = sum_pv_fcf + pv_exit_multiple_tv

print(f"\nSum of PV of explicit-period FCFs: EUR {sum_pv_fcf:,.0f}m")
print(f"PV of Gordon Growth terminal value: EUR {pv_gordon_growth_tv:,.0f}m")
print(f"PV of Exit Multiple terminal value: EUR {pv_exit_multiple_tv:,.0f}m")
print(f"\nEnterprise Value (Gordon Growth method): EUR {ev_gordon_growth:,.0f}m")
print(f"Enterprise Value (Exit Multiple method):  EUR {ev_exit_multiple:,.0f}m")

# ==============================================================================================
#  EV -> EQUITY VALUE -> PER-SHARE, FOR BOTH METHODS
# ==============================================================================================

# Long-term debt: Section 7 (itself sourced from Section 3, FY2025).
# Cash & cash equivalents: Section 3, FY2025 (includes short-term investments).
debt_eur_m = long_term_debt_eur_m
cash_eur_m = cash_flow_bs.loc["Cash & cash equivalents (incl. short-term investments)", "FY2025"]

# Net debt = Debt - Cash. ASML holds far more cash than debt, so this comes out NEGATIVE (net cash).
# Equity Value = EV - Net Debt handles the sign correctly on its own: subtracting a negative number
# adds the net cash back to EV, with no special-casing required.
net_debt_eur_m = debt_eur_m - cash_eur_m
print(f"\nLong-term debt (Section 7): EUR {debt_eur_m:,.1f}m")
print(f"Cash & equivalents (Section 3, FY2025): EUR {cash_eur_m:,.1f}m")
print(f"Net debt = Debt - Cash = EUR {net_debt_eur_m:,.1f}m "
      f"({'NET CASH position (negative net debt)' if net_debt_eur_m < 0 else 'net debt position'})")

diluted_shares_m = cash_flow_bs.loc["Weighted avg shares - diluted (millions)", "FY2025"]

equity_value_gordon = ev_gordon_growth - net_debt_eur_m
equity_value_exit = ev_exit_multiple - net_debt_eur_m

value_per_share_gordon = equity_value_gordon / diluted_shares_m
value_per_share_exit = equity_value_exit / diluted_shares_m

valuation_bridge = pd.DataFrame({
    "Gordon Growth method": {
        "Enterprise Value": f"€{ev_gordon_growth:,.0f}m",
        "less: Net debt": f"€{net_debt_eur_m:,.1f}m",
        "= Equity Value": f"€{equity_value_gordon:,.0f}m",
        "/ Diluted shares (m)": f"{diluted_shares_m:.1f}m",
        "= Value per share": f"€{value_per_share_gordon:,.2f}",
    },
    "Exit Multiple method": {
        "Enterprise Value": f"€{ev_exit_multiple:,.0f}m",
        "less: Net debt": f"€{net_debt_eur_m:,.1f}m",
        "= Equity Value": f"€{equity_value_exit:,.0f}m",
        "/ Diluted shares (m)": f"{diluted_shares_m:.1f}m",
        "= Value per share": f"€{value_per_share_exit:,.2f}",
    },
})
print()
display(valuation_bridge)

Present value of each explicit-period FCF (Section 6 FCF x discount factor at WACC):


,FY2026,FY2027,FY2028,FY2029,FY2030
Unlevered FCF,"€10,441m","€12,700m","€13,158m","€15,901m","€18,510m"
Discount period,1,2,3,4,5
Discount factor,0.9151,0.8375,0.7664,0.7014,0.6418
PV of FCF,"€9,555m","€10,636m","€10,085m","€11,152m","€11,880m"



Sum of PV of explicit-period FCFs: EUR 53,308m
PV of Gordon Growth terminal value: EUR 166,599m
PV of Exit Multiple terminal value: EUR 503,666m

Enterprise Value (Gordon Growth method): EUR 219,907m
Enterprise Value (Exit Multiple method):  EUR 556,973m

Long-term debt (Section 7): EUR 2,709.0m
Cash & equivalents (Section 3, FY2025): EUR 13,321.9m
Net debt = Debt - Cash = EUR -10,612.9m (NET CASH position (negative net debt))



,Gordon Growth method,Exit Multiple method
Enterprise Value,"€219,907m","€556,973m"
less: Net debt,"€-10,612.9m","€-10,612.9m"
= Equity Value,"€230,520m","€567,586m"
/ Diluted shares (m),388.9m,388.9m
= Value per share,€592.75,"€1,459.47"


### Comparison to current market price

ASML's actual traded share price already reflects the market's own blend of growth expectations,
risk, and everything else — comparing our two model-derived prices against it shows where this DCF's
range sits relative to what the market is currently paying, not a claim that the market is wrong.

In [10]:
# Current share price: Euronext Amsterdam listing (ticker ASML.AS), EUR.
# Source: Yahoo Finance (https://finance.yahoo.com/quote/ASML.AS/), intraday quote captured
# Sept 15 2026, 1:25:14 PM GMT+2, market open - EUR 1,406.80. This is a live/intraday price and
# will already be out of date by the time this notebook is re-run; re-check before relying on it.
current_share_price_eur = 1406.80

pct_vs_gordon = (value_per_share_gordon / current_share_price_eur) - 1
pct_vs_exit = (value_per_share_exit / current_share_price_eur) - 1

comparison = pd.DataFrame({
    "Gordon Growth method": {
        "Implied value per share": f"€{value_per_share_gordon:,.2f}",
        "Current share price": f"€{current_share_price_eur:,.2f}",
        "% above / (below) current price": f"{pct_vs_gordon:+.1%}",
    },
    "Exit Multiple method": {
        "Implied value per share": f"€{value_per_share_exit:,.2f}",
        "Current share price": f"€{current_share_price_eur:,.2f}",
        "% above / (below) current price": f"{pct_vs_exit:+.1%}",
    },
})
display(comparison)

low, high = sorted([value_per_share_gordon, value_per_share_exit])
print(f"\nImplied value range: EUR {low:,.2f} - EUR {high:,.2f} per share "
      f"(vs. current price of EUR {current_share_price_eur:,.2f}).")
if low <= current_share_price_eur <= high:
    print("The current market price falls INSIDE this range.")
elif current_share_price_eur < low:
    print("The current market price sits BELOW this entire range.")
else:
    print("The current market price sits ABOVE this entire range.")

,Gordon Growth method,Exit Multiple method
Implied value per share,€592.75,"€1,459.47"
Current share price,"€1,406.80","€1,406.80"
% above / (below) current price,-57.9%,+3.7%



Implied value range: EUR 592.75 - EUR 1,459.47 per share (vs. current price of EUR 1,406.80).
The current market price falls INSIDE this range.


## 10. Sensitivity Analysis

Section 9 produced two single-point implied values per share, each resting on a specific WACC and
either a specific terminal growth rate or a specific exit multiple. This section stress-tests both
of those single points by varying the two inputs each method is most sensitive to, holding
everything else fixed at Section 6's forecast and Section 9's net debt / share count — the same
full EV → Equity Value → per-share formula as Section 9, just re-run across a grid of WACC and
terminal-value inputs instead of at one base case.

**Color coding:** green cells imply a value *above* the current market price (€1,406.80, Section 9);
red cells imply a value *below* it. **The bordered cell in each grid is the base case** — the same
WACC (~9%, actually 9.27%) and terminal assumption (2.0% growth / 29.06x multiple) used in Sections
8–9 — so you can see at a glance how much the single-point answer moves if WACC or the terminal
assumption is even modestly different from what was assumed there.

### Grid 1 — Gordon Growth method

Implied value per share as WACC (rows) and the Gordon Growth terminal growth rate (columns) vary.
This is the more conservative of the two methods (Section 8/9) — **every cell in this grid is red**:
even the most favorable combination tested (WACC 7%, terminal growth 3.0%) only implies €1,047 per
share, still 25.6% below the current €1,406.80 market price. The Gordon Growth method can't get
close to the current market price anywhere in this parameter range, which is itself informative
(see the summary below).

In [11]:
def implied_value_per_share_gordon(wacc_input, terminal_growth_input):
    """Same EV -> Equity Value -> per-share formula as Section 9's Gordon Growth method, with
    WACC and terminal growth rate as free parameters. Reuses Section 6's forecast FCF and
    Section 9's net debt / diluted share count - nothing else changes."""
    periods = pd.Series(range(1, len(forecast.columns) + 1), index=forecast.columns)
    discount_factors_local = 1 / (1 + wacc_input) ** periods
    pv_fcf_local = (forecast.loc["Unlevered Free Cash Flow"] * discount_factors_local).sum()
    tv_local = (
        forecast.loc["Unlevered Free Cash Flow", "FY2030"]
        * (1 + terminal_growth_input)
        / (wacc_input - terminal_growth_input)
    )
    pv_tv_local = tv_local * discount_factors_local["FY2030"]
    ev_local = pv_fcf_local + pv_tv_local
    equity_value_local = ev_local - net_debt_eur_m
    return equity_value_local / diluted_shares_m


wacc_grid_values = [0.07, 0.08, 0.09, 0.10, 0.11, 0.12]
terminal_growth_grid_values = [0.005, 0.010, 0.015, 0.020, 0.025, 0.030]

wacc_row_labels = [f"{w:.0%}" for w in wacc_grid_values]
terminal_growth_col_labels = [f"{g:.1%}" for g in terminal_growth_grid_values]

gordon_grid = pd.DataFrame(index=wacc_row_labels, columns=terminal_growth_col_labels, dtype=float)
for w, w_label in zip(wacc_grid_values, wacc_row_labels):
    for g, g_label in zip(terminal_growth_grid_values, terminal_growth_col_labels):
        gordon_grid.loc[w_label, g_label] = implied_value_per_share_gordon(w, g)

print("Gordon Growth grid - raw values (EUR per share):")
display(gordon_grid.map(lambda v: f"{v:,.0f}"))

# Base case: WACC 9.27% (Section 7) rounds to the "9%" row; terminal growth 2.0% (Section 8) is the
# "2.0%" column exactly.
gordon_base_row = "9%"
gordon_base_col = "2.0%"


def highlight_vs_market(v):
    color = "#c6efce" if v > current_share_price_eur else "#ffc7ce"  # pastel green / red
    return f"background-color: {color}"


gordon_styled = (
    gordon_grid.style
    .map(highlight_vs_market)
    .format("€{:,.0f}")
    .set_properties(
        subset=pd.IndexSlice[[gordon_base_row], [gordon_base_col]],
        **{"border": "3px solid black", "font-weight": "bold"},
    )
    .set_caption("Gordon Growth: implied EUR value per share (rows = WACC, columns = terminal growth rate)")
)
display(gordon_styled)

Gordon Growth grid - raw values (EUR per share):


,0.5%,1.0%,1.5%,2.0%,2.5%,3.0%
7%,698,745,800,866,947,"1,047"
8%,604,637,675,720,773,837
9%,531,556,584,616,653,696
10%,474,493,514,538,565,596
11%,428,443,460,478,499,522
12%,390,402,415,430,446,463


,0.5%,1.0%,1.5%,2.0%,2.5%,3.0%
7%,€698,€745,€800,€866,€947,"€1,047"
8%,€604,€637,€675,€720,€773,€837
9%,€531,€556,€584,€616,€653,€696
10%,€474,€493,€514,€538,€565,€596
11%,€428,€443,€460,€478,€499,€522
12%,€390,€402,€415,€430,€446,€463


### Grid 2 — Exit Multiple method

Implied value per share as WACC (rows) and the exit EV/EBITDA multiple (columns) vary. This grid is
mostly green: the base-case cell (WACC ~9%, 29.06x) and everything to its right (higher multiples)
or above it (lower WACC) clears the current market price. The red pocket sits at low multiples
combined with higher WACC — the 20x column is red at every WACC tested, and 25x turns red once WACC
reaches 8% or higher.

In [12]:
def implied_value_per_share_exit(wacc_input, exit_multiple_input):
    """Same EV -> Equity Value -> per-share formula as Section 9's Exit Multiple method, with WACC
    and the exit EV/EBITDA multiple as free parameters. Reuses Section 6's forecast FCF/EBITDA and
    Section 9's net debt / diluted share count - nothing else changes."""
    periods = pd.Series(range(1, len(forecast.columns) + 1), index=forecast.columns)
    discount_factors_local = 1 / (1 + wacc_input) ** periods
    pv_fcf_local = (forecast.loc["Unlevered Free Cash Flow"] * discount_factors_local).sum()
    tv_local = fy2030_ebitda * exit_multiple_input
    pv_tv_local = tv_local * discount_factors_local["FY2030"]
    ev_local = pv_fcf_local + pv_tv_local
    equity_value_local = ev_local - net_debt_eur_m
    return equity_value_local / diluted_shares_m


exit_multiple_grid_values = [20, 25, 29.06, 35, 40]
exit_multiple_col_labels = [f"{m:g}x" for m in exit_multiple_grid_values]

exit_grid = pd.DataFrame(index=wacc_row_labels, columns=exit_multiple_col_labels, dtype=float)
for w, w_label in zip(wacc_grid_values, wacc_row_labels):
    for m, m_label in zip(exit_multiple_grid_values, exit_multiple_col_labels):
        exit_grid.loc[w_label, m_label] = implied_value_per_share_exit(w, m)

print("Exit Multiple grid - raw values (EUR per share):")
display(exit_grid.map(lambda v: f"{v:,.0f}"))

# Base case: WACC 9.27% (Section 7) rounds to the "9%" row; exit multiple 29.06x (Section 8) is the
# "29.06x" column exactly.
exit_base_row = "9%"
exit_base_col = "29.06x"

exit_styled = (
    exit_grid.style
    .map(highlight_vs_market)
    .format("€{:,.0f}")
    .set_properties(
        subset=pd.IndexSlice[[exit_base_row], [exit_base_col]],
        **{"border": "3px solid black", "font-weight": "bold"},
    )
    .set_caption("Exit Multiple: implied EUR value per share (rows = WACC, columns = exit EV/EBITDA multiple)")
)
display(exit_styled)

Exit Multiple grid - raw values (EUR per share):

,20x,25x,29.06x,35x,40x
7%,"1,164","1,411","1,612","1,906","2,154"
8%,"1,115","1,351","1,543","1,823","2,060"
9%,"1,068","1,294","1,477","1,745","1,971"
10%,"1,024","1,239","1,414","1,671","1,886"
11%,982,"1,188","1,355","1,600","1,806"
12%,942,"1,139","1,299","1,533","1,730"


,20x,25x,29.06x,35x,40x
7%,"€1,164","€1,411","€1,612","€1,906","€2,154"
8%,"€1,115","€1,351","€1,543","€1,823","€2,060"
9%,"€1,068","€1,294","€1,477","€1,745","€1,971"
10%,"€1,024","€1,239","€1,414","€1,671","€1,886"
11%,€982,"€1,188","€1,355","€1,600","€1,806"
12%,€942,"€1,139","€1,299","€1,533","€1,730"


### Where the current market price sits in each grid

- **Gordon Growth grid: every single cell is below the current market price.** Even the most
  favorable combination tested — WACC 7% (a point or more below Section 7's 9.27% base case) and
  3.0% terminal growth (well above the 2% ECB-inflation-anchored rate actually used) — only reaches
  €1,047, still 25.6% below €1,406.80. **The market price is not explainable by the Gordon Growth
  method anywhere in this grid**, not just at the base case — reinforcing, more starkly than Section
  9's single point (−57.9%) already suggested, that the market is not pricing ASML as a "matures to
  a stable, modest-growth business" story.
- **Exit Multiple grid: the current price sits comfortably inside the green region, not at its
  edge.** At the base-case WACC (~9%) and ASML's own 10-year median multiple (29.06x), the implied
  value (€1,477 in this grid; €1,459.47 exactly in Section 9) already clears the market price. The
  grid only turns red at low multiples (20x is red at every WACC tested; 25x turns red once WACC
  reaches 8% or higher) — both well below the *current* traded multiple of ~39–44x noted in Section
  8, which would push even further into green.
- **Taken together**, the two grids sharpen Section 9's core point rather than soften it: it isn't
  that the Gordon Growth base case is merely somewhat conservative — across this entire tested range
  it can't reach the current price at all. The market's valuation of ASML is really only explicable
  through something closer to the Exit Multiple method's implicit growth story.